02-KernelModules_CharacterDevice/CharDevSource.c

```diff
@@ -0,0 +1,269 @@
+	
+	sscanf(buffer, "%d\n", &clear_memory);
+	printk(KERN_INFO "ChD: clear_memory now = %d\n", clear_memory);
+	
+	if(clear_memory){
+		memset(data_buff, 0, (int) (sizeof data_buff));
+		kfree(data_buff);
+		data_buff = kzalloc(buffer_size+buffer_increase, GFP_KERNEL);
+		used_memory=0;
+		printk(KERN_INFO "ChD: Cleared memory\n");
+	}
+	return len;
+}
+
+static struct kobj_attribute sysfs_attribute =__ATTR(clear_memory, 0220, NULL, sysfs_store);
+
+//===/proc file operations===
+
+static int create_proc(void){
+    proc_dir = proc_mkdir(PROC_DIRECTORY, NULL);
+    if (NULL == proc_dir)
+        return -EFAULT;
+
+    proc_file = proc_create(PROC_FILENAME, S_IFREG | S_IRUGO | S_IWUGO, proc_dir, &proc_fops);
+    if (NULL == proc_file)
+        return -EFAULT;
+
+    return 0;
+}
+
+
+static void cleanup_proc(void){
+    if (proc_file)
+    {
+        remove_proc_entry(PROC_FILENAME, proc_dir);
+        proc_file = NULL;
+    }
+    if (proc_dir)
+    {
+        remove_proc_entry(PROC_DIRECTORY, NULL);
+        proc_dir = NULL;
+    }
+}
+
+static int proc_read(struct file *filp, char *buffer, size_t len, loff_t *offset ){
+	
+
+	int result;
+	char msg[124];
+	
+	sprintf(msg, "%d %d\n", buffer_size+buffer_increase, buffer_size+buffer_increase-used_memory);
+
+	
+	if(*offset >= strlen(msg)) {
+		*offset=0;
+		return 0;
+	}
+	
+	if(len > strlen(msg) - *offset)
+		len = strlen(msg) - *offset;
+
+	result = raw_copy_to_user((void*)buffer, msg - *offset, len);
+	
+	*offset += len;
+	
+	return len;
+}
+
+//===/dev file operations===
+
+int dev_open(struct inode *inode, struct file *filp){	
+	if(is_open){
+		printk(KERN_ERR "ChD: Failed to open - device is busy");
+		return -EBUSY;
+	}
+	is_open = 1;
+	return 0;
+}
+
+int dev_release(struct inode *inode, struct file *filp){
+	is_open = 0;
+	return 0;
+}
+
+ssize_t dev_read(struct file *filep, char *buffer, size_t len, loff_t *offset){
+	ssize_t retval = 0;
+
+	int remain = buffer_size - (int) (*offset);
+	if (remain == 0) {
+		goto out;
+	}
+
+	if (len > remain) {
+		len = remain;
+	}
+
+	if (raw_copy_to_user(buffer, data_buff + *offset, len)) {
+		retval = -EFAULT;
+		goto out;
+	}
+	*offset += len;
+	retval = len;
+
+	out:
+		return retval;
+}
+
+ssize_t dev_write(struct file *filep, const char *buffer, size_t len, loff_t *offset){
+
+	int error = 0;
+	int remain = buffer_size - (int) (*offset);
+	
+	used_memory += len;
+	
+	if(0 == remain)
+		return 0;
+	if(used_memory > remain)
+		return -EIO;
+	
+	error = raw_copy_from_user(data_buff + *offset, buffer, len);
+	
+	if(error){
+		printk(KERN_INFO "ChD: Failed to receive %zu characters from the user\n", len);
+		return -EFAULT;
+	}
+	else{
+		printk(KERN_INFO "ChD: Received %zu characters from the user\n", len);
+		offset += len;
+		return len;
+
+	}
+}
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
+
+//===init/exit functions===
+
+static int __init ChD_init(void){
```

Comment:
```suggestion
static int __init ChD_init(void)
{
```
Please fix  Linux Kernel Coding Style
However, there is one special case, namely functions: they have the opening brace at the 
beginning of the next line, thus:
int function(int x)
{
        body of function
}